### PR TITLE
Fix Logger in Mixin Usage

### PR DESCRIPTION
**Pull request by MarcelKoch**
This PR fixes logging used in the following mixins:
- EnablePolymorphicObject
- EnableDefaultLinOpFactory
Before, any class derived from EnablePolymorphicObject would not log the PolymorphicObject events. In the EnableDefaultLinOpFactory case, the generate function with logging is implemented in LinOpFactory, but in the current version this function does not get called. Instead, the default generate function is called (this has also been noted by @upsj).

This removes the adding of loggers from the AbstractFactory, so now only EnableDefaultFactory adds logger, if the concrete type supports it. And additionally, some implementation is simplified. Some tests are added to make sure that we actually log all events.

### CODE REVIEW

**Comment by upsj** [comment]
Seems like we have a bit of parallel work going on, I'm working on the same thing, though with a slightly different approach

File: core/test/base/executor.cpp [hunk 1]
```diff
@@ -651,4 +651,131 @@ TEST(ExecutorDeleter, AvoidsDeletionForNullExecutor)
...
+    }
+
+
+    void on_operation_completed(const gko::Executor* exec,
+                                const gko::Operation* op) const override
+    {
+        operation_completed++;
+    }
+
+    int mutable allocation_started = 0;
```
Comment (upsj) [inline core/test/base/executor.cpp:717]: nit: const west

File: core/test/base/executor.cpp [hunk 2]
```diff
@@ -651,4 +651,131 @@ TEST(ExecutorDeleter, AvoidsDeletionForNullExecutor)
...
+    int mutable alloctaion_completed = 0;
```
Comment (upsj) [inline core/test/base/executor.cpp:718]:
```suggestion
    int mutable allocation_completed = 0;
```

**Review by upsj — APPROVED** [review]
LGTM! Real nice job with the tests. I think it doesn't make sense to have a factory for a non-polymorphic type, so maybe we should keep logger transfer at the AbstractFactory level? Then we can also get rid of the propagate_logger setup.

**Comment by MarcelKoch** [comment]
This now contains the changes from #1038

**Review by keldu — APPROVED** [review]
LGTM

**Pull request merged**